### DISCUSSION TITLE
Native Vite Support

### DISCUSSION THREAD

**Discussion by Matelasse** [vercel/next.js#22406] [Help] [answered]
**Edit from @leerob:** Please see the answer below: https://github.com/vercel/next.js/discussions/22406#discussioncomment-1590537

**Comment by borekb** [comment]
This is a [great podcast episode about Vite](https://player.fm/series/series-1401837/ep-140-evan-you-reimagining-the-modern-dev-server-with-vite) with Evan You, got me thinking whether this could be married with Next.js 😄. Vite's approach sounded really quite "sound".

**Comment by rajington** [comment]
Thanks for sharing that podcast, would Vite's use of rollup maybe change how production is built as well? Or maybe something like Snowpack where it's just for dev? There sure are a lot of webpack-based plugins...

**Comment by Matelasse** [comment]
@Timer any thoughts regarding this feature request? i love both next.js and vite, and i would love to be able to use them both

**Comment by Janpot** [comment]
@Matelasse Did you see https://github.com/vercel/next.js/issues/15152 ?

**Comment by Matelasse** [comment]
@Janpot I hadn't seen that. Impressive and hopeful. Is there a way to experiment with wp5 before it's default in next?

**Comment by Matelasse** [comment]
update:
@Janpot webpack v5 is nowhere close to vite in terms of speed

**Comment by capaj** [comment]
why not switch to vite and drop webpack? That would be simpler to support in the future.

**Comment by zigomir** [comment]
How about now when it's [stable](https://github.com/vitejs/vite/releases/tag/v2.0.0)? :)

**Reply by spencerwooo** [reply]
Seems that the maintainers of Vite consider Next.js and itself to be separate build systems? And that they are either one or another? Don't know if this is true from Next.js' perspective. Discussion over at Vite is here: <https://github.com/vitejs/vite/issues/2156>

**Reply by soilSpoon** [reply]
> Seems that the maintainers of Vite consider Next.js and itself to be separate build systems? And that they are either one or another? Don't know if this is true from Next.js' perspective. Discussion over at Vite is here: [vitejs/vite#2156](https://github.com/vitejs/vite/issues/2156)

Since nextjs depends on webpack, it looks like nextjs can support vite, although vite can't support nextjs.

Edit: Also vite uses rollup and esbuild, and can only use esbuild in the future.

## Reference
https://vitejs.dev/guide/api-plugin.html
https://vitejs.dev/guide/why.html#why-not-bundle-with-esbuild

**Comment by brillout** [comment]
If you want to use Vite, check out [vite-plugin-ssr](https://github.com/brillout/vite-plugin-ssr).

It's a little SSR tool that has a similar interface than Next.js but built on top of Vite. It's tiny: only 1.3 kLOC(!) thanks to Vite's native SSR support. This means that it will soon become a rock-solid tool (there is nothing complex in its implementation, all nifty stuff is done by Vite).

Thanks to its simple design that has clear separation of concerns, thoughts like "Hm, I'm not sure how to integrate this tool with Next.js", "Hm, I'm not sure if I can achieve that with Next.js", or "I feel like Next.js is limiting me" will vanish. While you have slightly more boilerplate (only two files that define how your pages are rendered), you keep control over your stack in a simple and straighforward manner.

(Author here, if that wasn't obvious already :D.)

**Reply by truongsinh** [reply]
I checked it out, but I think what stopped me from adopting it, is the fact that it advocates server routing instead of client routing, that "routing" cannot be both, but must be either-or, and that developers have to distinguish *.page.server.js and *.page.client.js . To extend the context, routing in both NextJs and Sapper/SvelteKit are universal (work for both SSR and CSR).

I do understand the difficulty and complexity of doing it, but, IMHO, it's the reason to use frameworks in the first place.

**Reply by brillout** [reply]
@truongsinh

What are you building? Whether to use Server Routing or Client Routing really depends on the kind of app you are building.

> I do understand the difficulty and complexity of doing it, but, IMHO, it's the reason to use frameworks in the first place.

That's actually not the case. Next.js/Nuxt cannot make complexity magically disappear. It may not be apparent at first but you'll eventually pay the debt as you scale sophisticated apps.

Either way, `vite-plugin-ssr` has first-class support for Client Routing as well.

I revamped https://vite-plugin-ssr.com/SR-vs-CR. Let me know if something's not clear.

**Reply by truongsinh** [reply]
> What are you building? 

Interactive webpapp over unreliable/throttled network, but can use latest protocol, such as HTTP2/HTTP3.

>  It may not be apparent at first but you'll eventually pay the debt as you scale sophisticated apps.

Do you happen to have blog about that? I'd like to learn about this for many reasons, (a) if I still use Next/Nuxt, for the next mega-project, I'd rather learn the problem now rather than later, (b) whether it's a formal/mathematical problem that cannot be solved (NP-complete kind of thing), or just that we have not found the "algo" yet.

> I revamped https://vite-plugin-ssr.com/SR-vs-CR. Let me know if something's not clear.

Sorry still not clear to me, like with CR, can I use Filesystem Routing? in https://vite-plugin-ssr.com/useClientRouter 
```
// _default.page.client.js
// Environment: Browser
```

So this file `_default.page.client.js` can never be executed on NodeJS? And if we want to serve the same page, we have to duplicate the logic to another file _default.page.server.js?

**Reply by brillout** [reply]
> with CR, can I use Filesystem Routing?

Yes, see https://github.com/brillout/vite-plugin-ssr/tree/master/examples/react-full

> whether it's a formal/mathematical problem that cannot be solved

Yes there is inherent adedd complexity when using CR.

When the user navigates to a page, there are basically two scenarios with CR:
 1. It's the first page the user is visiting => HTML is loaded & hydrated.
 2. The user is already on a page => only the code of the new page is loaded and the DOM is mutated.

Now with SR, you never have scenario 2 and you always get scenario 1 instead.

This means your code (and your architecture at large) as one scenario less to consider. This means less code and less complexity.

For example, if you want to integrate GraphQL, then you'll have to consider these two scenarios when using CR. With SR you only have to consider scenario 1.

Now because Next.js is a black-box you don't realize the added complexity at first, but it will make your life more difficult as soon as you get off the Next.js beaten path.

Does that make sense?

**Comment by klarkc** [comment]
I am currently using vite with `Other` preset on vercel (without SSR), adding native vite support with ssr would greatly improve developer experience

**Comment by ritwikvd** [comment]
@timneutkens Is there any update whatsoever on this? You don't need me telling you that webpack for dev is torture. If Vite/ Snowpack can't be integrated into Next, what about a custom solution made in house? We need movement on this.

**Reply by borekb** [reply]
The author of webpack, @sokra, just joined Vercel: https://twitter.com/wSokra/status/1381860800533528576.

The Next.js team cares about DX a lot so I trust this is a great move. Congrats @sokra!

**Reply by redbar0n** [reply]
**How to use a bit of esbuild with NextJS today:**
@ritwikvd For a solution today, try [esbuild-loader-examples](https://github.com/privatenumber/esbuild-loader-examples/tree/master/examples/next). It shows how to use esbuild with NextJS, via [esbuild-loader](https://github.com/privatenumber/esbuild-loader) that lets you "harness the speed of esbuild in your Webpack build by offering faster alternatives for transpilation (eg. babel-loader/ts-loader) and minification (eg. Terser)". The [speedup gains you'd be able to make](https://github.com/privatenumber/esbuild-loader/discussions/138) are typically 2x, but some even report gains close to pure esbuild setups. (This is for build and not for dev/HMR).

See also: https://github.com/evanw/esbuild/issues/838#issuecomment-839908523

PS: I am not a maintainer, just strolling by, sharing what I've discovered.

**Comment by leerob** [comment]
Related: We will be open-sourcing the brand new dev server behind [Next.js Live](https://nextjs.org/live), which uses some exciting new tech to remove the build process entirely as Next can now run in the browser. As you can see from the [latency numbers](https://youtu.be/ze8ycxc1UzE?t=238), it's really fast 😮

**Reply by matepaiva** [reply]
I would love to have access to that!

**Reply by truongsinh** [reply]
The latency numbers are indeed impressive. However, as so many have pointed out already, the core difference between Webpack vs Vitejs is bundle vs unbundle. That is, Vitejs has constant O(1) time complexity w.r.t the number of relevant files (JS/TS, CSS, HTML, etc), while Webpack has O(n).

Put it another way, with the demo, assuming the number of relevant files in the project is 1e3, latency is sub-second. My question is, when the number of relevant files scales to 1e6, can the latency remain sub-second?

(btw, feel free to invalidate my assumption about projects may scale to 1e6 files, possibly with sth like "you should re-organize your code to module federation when you reach that scale")

**Reply by truongsinh** [reply]
@leerob , because I got no answers, I created a dummy project myself https://github.com/truongsinh/webpack5-is-slow-example . Here, I use the latest Next (11) and webpack (5), there are at least 10k (1e4) js file, each with a simple line `export const value = 42;`. Cold start-up is like 2 minutes, and rebuild is like 10 seconds. Originally I intend to try 1e6, but I think the point is clear.

**Reply by shuding** [reply]
@truongsinh Next.js Live doesn’t have bundling process. It’s based on native ESM: https://twitter.com/shuding_/status/1405121713629806592.

**Reply by truongsinh** [reply]
Fantastic, that's exactly the answer I was looking for. Can't wait to try `Next.js Live` on 10k-file project. I still have concerns about how service workers consumed resources to track multiple files (and file changes), but highly probably it's solvable problem.

**Reply by truongsinh** [reply]
@shuding I just learn that I can already try Next.js Live today by cloning and deploying the commerce kit. However, when I follow the instruction, the code seems to be broken https://commerce-byterow6e-truongsinh-3.vercel.live
<img width="945" alt="Screen Shot 2021-09-04 at 9 45 26 AM" src="https://user-images.githubusercontent.com/358585/132098467-6781a823-b62f-49c5-95b9-b6611efc00c8.png">

**Comment by s-kris** [comment]
@leerob Any tentative timeline for the release? 
Thank you!

**Comment by truongsinh** [comment]
I created a dummy project myself https://github.com/truongsinh/webpack5-is-slow-example . Here, I use the latest Next (11) and webpack (5), there are at least 10k (1e4) js files, each with a simple line `export const value = 42;`. Cold start-up is like **2 minutes**, and rebuild is like **10 seconds**. Originally I intend to try 1e6, but I think the point is clear.

So, apparently, the problem is still O(n) vs O(1). I'm excited to see whether Next.js Live can overcome this. Apparently, 10k-file project is not uncommon, there are members complaining about 15k-file project (and thus had to abandon NextJS and move to Vite + Vite SSR). To give more context, the codebase (admittedly with tech debt) in my day job has 130k files (10 minutes cold build and 90 seconds incremental rebuild, on shared dev machine Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz 16-core)

**Reply by sokra** [reply]
Are you using windows?

You could try to add that to the next.config.js, which should lead to faster builds in for that repo:

``` js
  webpack: (config, { isServer }) => {
    // faster rebuilds for many files:
    config.module.unsafeCache = true;
    // workarounds for windows:
    config.plugins.push((c) => (c.options.devtool = 'eval-source-map'));
    return config;
  },
```

Also try to disable anti-vir

**Reply by truongsinh** [reply]
> Are you using windows?

no, I'm using Mac (locally) or Linux (remote dev machine).

> You could try to add that to the next.config.js, which should lead to faster builds in for that repo:

so, I will use only "config.module.unsafeCache = true;" ? What kind of "unsafe" are we talking about?

> Also try to disable anti-vir

I don't think I am either able or willing to.

I mean, the repo is public, did you have different results when try running webpack on it?

**Reply by timneutkens** [reply]
Thanks for your feedback!

I've ran some comparisons against that repo, including Vite's server-side usage:
- When using Vite compiling `pages/index.js` with JSX breaks completely so I removed the JSX
  - Compile on my machine took between 16 and 18 seconds for the single-pass compile, so I'll compare that with a single-pass compile (only client-side compilation) of Next.js (Next.js does 2, one for client one for server, and to mirror Next.js in Vite you'd have to do 2 as well to have the correct output)
- When using Next.js latest canary using Babel (output es5)
  - Compile on my machine took between 51 and 55 seconds for a single-pass compile (logical, as all files in the project are processed through Babel)
- When using Next.js latest canary using SWC (output es5)
  - Compile on my machine took between 18 and 20 seconds for a single-pass compile, however note that this is compiling output to es5 for backwards compatibility, we'll also have a flag for full esmodules output which will likely be even faster

The main thing to note here is that *recompilation* (Fast Refresh) with a change on `pages/index.js` takes about 3 milliseconds when using SWC, the rest of the time spend, about 2,1 second is spent writing to disk as the output file is **41MB** (do note this would be a severe outlier from what we see in general for a single JS bundle).

There's a few action items that we are working on already:
- Moving all compilation to SWC, this is very close to a public beta, we're currently finishing up the styled-jsx transform as the SWC support will be fully compatible with existing Next.js applications that do not have custom Babel configuration
- Adding an ESModules output target that does fewer transforms
- Not writing to disk in development but instead keeping the files in-memory (will get rid of that writing overhead)
  - This will also solve an issue on Windows where we use `inline-source-map` instead of `eval-source-map` because of anti-virus (which is why @sokra asked if you're on Windows)
- Leveraging webpack 5 on-demand compilation of `import()`. We already have on-demand compilation of entries (pages). This requires removing webpack 4 support (will be done in Next.js 12)
- Hibernating the server-side compilation when there's no request to server-side render a page. Currently client-side compilation is already prioritized, this would just improve it a small bit more.

The main thing that is often a misconception is that webpack will recompile a full bundle (e.g. run SWC or Babel over all deps of a file that is changed) and then send that complete bundle to the browser. Both are not the case, only the exact file that you change (e.g. `components/button.js`) will be recompiled. The module change will then be transmitted through an eventsource connection and then the delta will be fetched + applied in the browser.


Note: the measurements I took were all on **cold** builds, removing `.next`, so no webpack caching was leveraged. When webpack caching is leveraged (enabled by default) rebooting / opening a page is significantly faster (as it's cached already) whereas rebooting the Vite server-side example does a full recompilation of all files (the 16-18s)

**Comment by truongsinh** [comment]
Thanks for the explanation

> Vite server-side example does a full recompilation of all files (the 16-18s)

From developer experience, they don't perceive "full compilation", they perceive "first load" (similar to TTFB), and Vite waits until the first time a file is requested to compile, thus I'd say this info is kind-of irrelevant. One can say that Vite "cheat" by having the dev server up instantly, and TTFB is almost 0, but `time to first paint` can still be affected by N (number of files). Does your "18 seconds" refer to `time to first paint`? Also, talking about cache, I believe (please don't assume I'm right on this) Vite also utilizes cache, is "full compilation" should not take longer than necessary when comparing webpack's cache with vite's cache.


> the rest of the time spend, about 2,1 second is spent writing

> The main thing that is often a misconception is that webpack will recompile a full bundle

I don't know who has that misconception, but I think it's quite straightforward. The problem (or could be my misconception) is that even only file A changes in the set of file A, B, C, D (and thus B, C, D does not need to recompile), the "concatenation" still needs to happens, and thus needs both read and write to all A, B, C, D. That is, again, the main difference between bundled and unbundled, or O(1) vs O(n). For webpack, I would assume it's the IO of reading all files (did I hear something about moving this to in-mem), the CPU of concatenating, and the writing of the new bundle (did I hear something about moving this to in-mem again).

Anyway, as a (still potential) user of Next.JS
>   the rest of the time spend, about 2,1 

is there any way we get this to sub-second, that is the [latency](https://youtu.be/ze8ycxc1UzE?t=238) remains sub-second regardless of how the files are structured? As said, it's already possible with Vite. **One idea is to send the changed part as soon as it is re-compiled (but not wait for the concatenation)**.

> would be a severe outlier from what we see in general for a single JS bundle)

could be, like 99.9% of JS bundle would not be like this, but as mentioned, the rest 0.1% are the largest ones, usually those of big corps (potentially with legacy codebase). I'd totally understand if the business decision is not to support that 0.1%, but if that's the case, it would be very sad for me, because I would continue to use Next.js for my side project, but would not be able to recommend it for any serious full-time-job project (again see my other comments, where people just have to ditch NextJS when they reach 15k files)

**Reply by timneutkens** [reply]
> From developer experience, they don't perceive "full compilation", they perceive "first load" (similar to TTFB), and Vite waits until the first time a file is requested to compile, thus I'd say this info is kind-of irrelevant. One can say that Vite "cheat" by having the dev server up instantly, and TTFB is almost 0, but time to first paint can still be affected by N (number of files). Does your "18 seconds" refer to time to first paint? Also, talking about cache, I believe (please don't assume I'm right on this) Vite also utilizes cache, is "full compilation" should not take longer than necessary when comparing webpack's cache with vite's cache.

18 seconds refers to time to compilation complete with Vite's SSR API. Next.js does on-demand compilation and does not (like many other webpack-based applications) inject all entrypoints the moment you boot up `next dev`. It compiles entrypoints on-demand when you request them, same as Vite. As far as I've investigated Vite does not have caching enabled by default (maybe it relies on the browser cache, but that is not a thing in Node.js compilation), every reboot it would take 18s to get something on the screen. Do note that it then still needs a client-side compilation as well to correctly compare to Next.js' compilation output (SSR vs client differ as explained). The main reason you see Vite boot up instantly is because is does practically nothing by default, in case of Next.js it starts compiling files immediately after binding the http server to ensure files that we know you're going to need are compiled even before you request the first page.

> I don't know who has that misconception, but I think it's quite straightforward. The problem (or could be my misconception) is that even only file A changes in the set of file A, B, C, D (and thus B, C, D does not need to recompile), the "concatenation" still needs to happens, and thus needs both read and write to all A, B, C, D. That is, again, the main difference between bundled and unbundled, or O(1) vs O(n). For webpack, I would assume it's the IO of reading all files (did I hear something about moving this to in-mem), the CPU of concatenating, and the writing of the new bundle (did I hear something about moving this to in-mem again).

As I explained in my previous comment this is exactly the misconception I was talking about. There is no reading of all files, as all files are already compiled in-memory and they are not changed. `B, C, D` would not have to be read from disk, it already reads that from cache. The main thing that I already highlighted is that concatenating (which is fast enough) and writing to disk of the result (which is slow) can be opted out of completely and that is something we're working on currently. For Fast Refresh webpack already only applies a delta of the specific module that was changed. So only `A` would be propagated to the browser. 

> is there any way we get this to sub-second, that is the latency remains sub-second regardless of how the files are structured? As said, it's already possible with Vite. One idea is to send the changed part as soon as it is re-compiled (but not wait for the concatenation).

As per the explanation I added to the previous question the `2.1s` of writing is something we're getting rid of entirely and that is currently in progress by instead of writing to disk writing to memory.

> could be, like 99.9% of JS bundle would not be like this, but as mentioned, the rest 0.1% are the largest ones, usually those of big corps (potentially with legacy codebase). I'd totally understand if the business decision is not to support that 0.1%, but if that's the case, it would be very sad for me, because I would continue to use Next.js for my side project, but would not be able to recommend it for any serious full-time-job project (again see my other comments, where people just have to ditch NextJS when they reach 15k files)

It seems that you misunderstood my comment. All the action items that I shared are related to making compilation significantly faster for both large and small apps while retaining compatibility with the vast majority of existing Next.js application that will just be able to `yarn add next@latest` to leverage these optimizations.

**Reply by truongsinh** [reply]
Thanks for taking the time to clarify. I'd say I'm not yet totally convinced, but the only thing matters (to me) is

> As per the explanation I added to the previous question the `2.1s` of writing is something we're getting rid of entirely and that is currently in progress by instead of writing to disk writing to memory.

As soon as this is available to be tested, I believe the asks for support for Vite/Snowpack/other unbundled builder will be much less relevant (whether it's totally irrelevant depends on how the developer experience is at that point :D). So, it goes without saying, I'm excitedly looking forward to it 😃

**Comment by filipesmedeiros** [comment]
i just want to chip in that all these updates that @timneutkens is sharing make me excited for the future of next and being a web dev in general! and i already love next today 😄

**Comment by craftogrammer** [comment]
Do we still need this in Next12, as it comes with Rust compiler? 😬

**Comment by leerob** [comment, marked as answer]
Hey everyone! With Next.js 12, we've adopted a new Rust compiler. Some quick numbers:

- 3x faster refresh
- 5x faster builds for production
- 17x faster than Babel for compilation
- 7x faster than Terser (for opt-in minification)

This is just the beginning and we will be continuing to invest here!

Read more: https://nextjs.org/12

**Reply by leerob** [reply]
To be extremely clear: by adopting Rust, we are not moving forward with native Vite support. I'm going to go ahead and close this discussion 🙏
